### PR TITLE
Fix/non null relationships

### PR DESCRIPTION
**Pull request by danstarns**
# Description

Enforces that Nodes have relationships, on update and create, if the field is annotated with an `!`. This PR uses `apoc.util.validate` to throw an error inside a transaction. 

# Issue

#549

### CODE REVIEW

File: packages/graphql/src/translate/create-create-and-params.ts [hunk 1]
```diff
@@ -245,6 +246,12 @@ function createCreateAndParams({
         creates.push(`CALL apoc.util.validate(NOT(${meta.authStrs.join(" AND ")}), ${forbiddenString}, [0])`);
     }
 
+    let relationshipValidationStr = createRelationshipValidationStr({ node, context, varName });
```
Comment (angrykoala) [inline packages/graphql/src/translate/create-create-and-params.ts:249]: I'm not sure why this is not const?

File: packages/graphql/src/translate/create-relationship-validation-str.ts [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+        const isArrayTypeRequired = field.typeMeta.arrayTypeRequired;
+        const isArray = field.typeMeta.array;
+        const isNotUnionOrInterface = Boolean(field.union) || Boolean(field.interface);
+
+        if (isArray) {
+            return isArrayTypeRequired && !isNotUnionOrInterface;
+        }
+
+        return isRequired && !isNotUnionOrInterface;
+    });
```
Comment (angrykoala) [inline packages/graphql/src/translate/create-relationship-validation-str.ts:27]: Consider moving this to a separate function: 
```suggestion
    const nonNullRelationFields = filterNonNullFields(node.relationFields);
```
(This is a suggestion, I'm happy to keep it as is too)

File: packages/graphql/src/translate/create-relationship-validation-str.ts [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+            const relTypeStr = `[:${field.type}]`;
+            const toNode = context.neoSchema.nodes.find((n) => n.name === field.typeMeta.name) as Node;
+            const exitsStr = `EXISTS((${varName})${inStr}${relTypeStr}${outStr}(${toNode.getLabelString(context)}))`;
+
+            return `apoc.util.validatePredicate(NOT(${exitsStr}), '${RELATIONSHIP_REQUIREMENT_PREFIX}${node.name}.${field.fieldName} required', [0])`;
+        });
+
+        relationshipValidationStr = `CALL apoc.util.validate(NOT(${nonNullPredicates.join(
+            " AND "
+        )}), '${RELATIONSHIP_REQUIREMENT_PREFIX}', [0])`;
```
Comment (angrykoala) [inline packages/graphql/src/translate/create-relationship-validation-str.ts:42]: Thoughts on just returning this?

```suggestion
       return `CALL apoc.util.validate(NOT(${nonNullPredicates.join(
            " AND "
        )}), '${RELATIONSHIP_REQUIREMENT_PREFIX}', [0])`;
```

This way we could get rid of the variable `relationshipValidationStr` and just return empty string at the end

File: packages/graphql/src/translate/create-relationship-validation-str.ts [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+    if (nonNullRelationFields.length) {
+        const nonNullPredicates = nonNullRelationFields.map((field) => {
+            const inStr = field.direction === "IN" ? "<-" : "-";
+            const outStr = field.direction === "OUT" ? "->" : "-";
+            const relTypeStr = `[:${field.type}]`;
+            const toNode = context.neoSchema.nodes.find((n) => n.name === field.typeMeta.name) as Node;
+            const exitsStr = `EXISTS((${varName})${inStr}${relTypeStr}${outStr}(${toNode.getLabelString(context)}))`;
+
+            return `apoc.util.validatePredicate(NOT(${exitsStr}), '${RELATIONSHIP_REQUIREMENT_PREFIX}${node.name}.${field.fieldName} required', [0])`;
+        });
```
Comment (angrykoala) [inline packages/graphql/src/translate/create-relationship-validation-str.ts:38]: Again, just an optional suggestion, but maybe moving this to a separate function could give a bit of extra clarity and context :thinking: :

```suggestion
        const nonNullPredicates = nonNullRelationFields.map((field) => {
            return createRelationFieldPredicate({field, context, node});
         });
```

WIth the separate function:

```ts
createRelationFieldPredicate({field, context, node}): string{
            const inStr = field.direction === "IN" ? "<-" : "-";
            const outStr = field.direction === "OUT" ? "->" : "-";
            const relTypeStr = `[:${field.type}]`;
            const toNode = context.neoSchema.nodes.find((n) => n.name === field.typeMeta.name) as Node;
            const exitsStr = `EXISTS((${varName})${inStr}${relTypeStr}${outStr}(${toNode.getLabelString(context)}))`;

            return `apoc.util.validatePredicate(NOT(${exitsStr}), '${RELATIONSHIP_REQUIREMENT_PREFIX}${node.name}.${field.fieldName} required', [0])`;

}

```

File: packages/graphql/tests/tck/tck-test-files/connections/projections/update.test.ts [hunk 5]
```diff
@@ -79,6 +79,8 @@ describe("Cypher -> Connections -> Projections -> Update", () => {
             "MATCH (this:Movie)
             WHERE this.title = $this_title
             WITH this
+            WITH this
```
Comment (angrykoala) [inline packages/graphql/tests/tck/tck-test-files/connections/projections/update.test.ts:82]: Having 2 consecutive `WITH this` is ok?
I feel we should try to avoid that, if its easy to do so

**Review by angrykoala — APPROVED** [review]
Approved! :+1:  just a few style suggestions.

File: packages/graphql/src/translate/create-create-and-params.ts (hunk 1 above)
Comment (danstarns) [reply packages/graphql/src/translate/create-create-and-params.ts]: Good call

File: packages/graphql/src/translate/create-relationship-validation-str.ts (hunk 2 above)
Comment (danstarns) [reply packages/graphql/src/translate/create-relationship-validation-str.ts:27]: I'm going to keep this as it is. I feel the function is pretty small in size and breaking it down will just add unnecessary context switching for future readers.

File: packages/graphql/src/translate/create-relationship-validation-str.ts (hunk 3 above)
Comment (danstarns) [reply packages/graphql/src/translate/create-relationship-validation-str.ts:42]: This is to ensure eslint does not complain about the consistent return.

File: packages/graphql/tests/tck/tck-test-files/connections/projections/update.test.ts (hunk 5 above)
Comment (danstarns) [reply packages/graphql/tests/tck/tck-test-files/connections/projections/update.test.ts:82]: This is an edge case that needs to be there, it wont effect the query in any way and is documented here https://github.com/neo4j/graphql/blob/master/packages/graphql/src/translate/translate-update.ts#L391
